pageserver/src/http/routes.rs

```diff
@@ -225,28 +249,86 @@ async fn timeline_attach_handler(request: Request<Body>) -> Result<Response<Body
     .await
     .map_err(ApiError::from_err)??;
 
-    let mut remote_index_write = get_state(&request).remote_index.write().await;
-
-    let _enter = span.entered(); // entered guard cannot live across awaits (non Send)
-    let index_entry = remote_index_write
-        .timeline_entry_mut(&ZTenantTimelineId {
-            tenant_id,
-            timeline_id,
-        })
-        .ok_or_else(|| ApiError::NotFound("Unknown remote timeline".to_string()))?;
+    let sync_id = ZTenantTimelineId {
+        tenant_id,
+        timeline_id,
+    };
+    let state = get_state(&request);
+    let remote_index = &state.remote_index;
+
+    let existing_remote_timeline = { remote_index.read().await.timeline_entry(&sync_id).cloned() };
+    match existing_remote_timeline {
+        Some(remote_timeline) => {
+            if remote_timeline.awaits_download {
+                return Err(ApiError::Conflict(
+                    "Timeline download is already in progress".to_string(),
+                ));
+            }
 
-    if index_entry.get_awaits_download() {
-        return Err(ApiError::Conflict(
-            "Timeline download is already in progress".to_string(),
-        ));
+            match remote_index.write().await.timeline_entry_mut(&sync_id) {
+                Some(remote_timeline) => remote_timeline.awaits_download = true,
+                None => {
+                    return Err(ApiError::NotFound(
+                        "Remote timeline got removed from the index during the operation, cannot proceed".to_string(),
+                    ))
+                }
+            };
+        }
+        None => match try_download_shard_data(state, sync_id).await {
+            Ok(Some(mut new_timeline)) => {
+                tokio::fs::create_dir_all(state.conf.timeline_path(&timeline_id, &tenant_id))
+                    .await
+                    .context("Failed to create new timeline directory")?;
+
+                new_timeline.awaits_download = true;
+                remote_index
+                    .write()
+                    .await
+                    .add_timeline_entry(sync_id, new_timeline);
+            }
+            Ok(None) => return Err(ApiError::NotFound("Unknown remote timeline".to_string())),
+            Err(e) => {
+                // TODO add a better way to detect "missing timeline" errors
+                error!("Failed to retrieve remote timeline data: {:?}", e);
+                return Err(ApiError::NotFound("Unknown remote timeline".to_string()));
+            }
+        },
     }
 
-    index_entry.set_awaits_download(true);
-    schedule_timeline_download(tenant_id, timeline_id);
+    let _enter = span.entered(); // entered guard cannot live across awaits (non Send)
```

Comment:
This span looks rather pathetic now, so I wonder, should we replace it with `.instrument()` on a higher level or remove entirely?